File: python/fred/fred_function.py
from __future__ import annotations

import io
import os
from typing import Optional

import pandas as pd
import requests


def fredgraph_api_function(graph_id: str) -> pd.DataFrame:
    url = "https://fred.stlouisfed.org/graph/fredgraph.csv"
    resp = requests.get(url, params={"g": graph_id}, headers={"Accept": "text/csv"}, timeout=120)
    resp.raise_for_status()
    text_stream = io.StringIO(resp.content.decode("utf-8"))
    return pd.read_csv(text_stream)


def fred_api_function(
    series_id: str,
    observation_start: Optional[str] = None,
    observation_end: Optional[str] = None,
    realtime_start: Optional[str] = None,
    realtime_end: Optional[str] = None,
    limit: Optional[int] = None,
    offset: Optional[int] = None,
    sort_order: str = "asc",
    units: Optional[str] = None,
    frequency: Optional[str] = None,
    aggregation_method: Optional[str] = None,
    output_type: Optional[int] = None,
    vintage_dates: Optional[str] = None,
    api_key: Optional[str] = None,
):
    if api_key is None:
        api_key = os.environ.get("FRED_API_KEY")
    if not api_key:
        raise RuntimeError("Defina la variable de entorno FRED_API_KEY para usar fred_api_function.")

    if units is not None:
        allowed_units = {"lin","chg","ch1","pch","pc1","pca","cch","cca","log"}
        if units not in allowed_units:
            raise ValueError(f"Valor de 'units' no válido: {units}")
    if frequency is not None:
        allowed_freq = {"d","w","bw","m","q","sa","a","wef","weth","wew","wetu","wem","wesu","wesa","bwew","bwem"}
        if frequency not in allowed_freq:
            raise ValueError(f"Valor de 'frequency' no válido: {frequency}")
    if aggregation_method is not None:
        allowed_agg = {"avg","sum","eop"}
        if aggregation_method not in allowed_agg:
            raise ValueError(f"Valor de 'aggregation_method' no válido: {aggregation_method}")
    if output_type is not None and output_type not in {1,2,3,4}:
        raise ValueError("'output_type' debe ser uno de 1,2,3,4")

    url = "https://api.stlouisfed.org/fred/series/observations"
    q = {
        "series_id": series_id,
        "api_key": api_key,
        "file_type": "json",
    }
    if observation_start: q["observation_start"] = observation_start
    if observation_end: q["observation_end"] = observation_end
    if realtime_start: q["realtime_start"] = realtime_start
    if realtime_end: q["realtime_end"] = realtime_end
    if limit is not None: q["limit"] = limit
    if offset is not None: q["offset"] = offset
    if sort_order: q["sort_order"] = sort_order
    if units: q["units"] = units
    if frequency: q["frequency"] = frequency
    if aggregation_method: q["aggregation_method"] = aggregation_method
    if output_type is not None: q["output_type"] = output_type
    if vintage_dates: q["vintage_dates"] = vintage_dates

    resp = requests.get(url, params=q, headers={"Accept": "application/json"}, timeout=120)
    resp.raise_for_status()
    obj = resp.json()
    if "observations" not in obj:
        raise RuntimeError("Estructura inesperada de la API de FRED; falta 'observations'.")
    return pd.DataFrame(obj["observations"])  # already flattened keys

File: R/fred/fred_function.R
# Funciones para descargar datos de FRED

# 1) Método fredgraph (no requiere API key) -> CSV directo de un gráfico compartido
fredgraph_api_function <- function(graphId) {
  url <- "https://fred.stlouisfed.org/graph/fredgraph.csv"
  res <- httr::GET(url, query = list(g = graphId), httr::accept("text/csv"))
  if (httr::http_error(res)) stop(sprintf("FRED fredgraph request failed [%s]", httr::status_code(res)))
  txt <- httr::content(res, as = "text", encoding = "UTF-8")
  df <- utils::read.csv(text = txt, stringsAsFactors = FALSE)
  return(df)
}

# 2) Método API v1 (requiere variable de entorno FRED_API_KEY) -> JSON con observaciones
fred_api_function <- function(
  series_id,
  observation_start = NULL,
  observation_end = NULL,
  realtime_start = NULL,
  realtime_end = NULL,
  limit = NULL,
  offset = NULL,
  sort_order = c("asc", "desc"),
  units = NULL,
  frequency = NULL,
  aggregation_method = NULL,
  output_type = NULL,
  vintage_dates = NULL,
  api_key = Sys.getenv("FRED_API_KEY", unset = NA_character_)
) {
  if (is.na(api_key) || api_key == "") stop("Defina la variable de entorno FRED_API_KEY para usar fred_api_function.")

  # Validaciones ligeras según documentación oficial
  if (!is.null(units)) {
    allowed_units <- c("lin", "chg", "ch1", "pch", "pc1", "pca", "cch", "cca", "log")
    if (!units %in% allowed_units) stop(sprintf("Valor de 'units' no válido: %s", units))
  }
  if (!is.null(frequency)) {
    allowed_freq <- c("d","w","bw","m","q","sa","a","wef","weth","wew","wetu","wem","wesu","wesa","bwew","bwem")
    if (!frequency %in% allowed_freq) stop(sprintf("Valor de 'frequency' no válido: %s", frequency))
  }
  if (!is.null(aggregation_method)) {
    allowed_agg <- c("avg","sum","eop")
    if (!aggregation_method %in% allowed_agg) stop(sprintf("Valor de 'aggregation_method' no válido: %s", aggregation_method))
  }
  if (!is.null(output_type)) {
    output_type <- as.integer(output_type)
    if (!output_type %in% c(1L,2L,3L,4L)) stop(sprintf("Valor de 'output_type' no válido: %s", output_type))
  }
  sort_order <- match.arg(sort_order)

  url <- "https://api.stlouisfed.org/fred/series/observations"
  q <- list(
    series_id = series_id,
    api_key = api_key,
    file_type = "json"
  )
  if (!is.null(observation_start)) q$observation_start <- observation_start
  if (!is.null(observation_end)) q$observation_end <- observation_end
  if (!is.null(realtime_start)) q$realtime_start <- realtime_start
  if (!is.null(realtime_end)) q$realtime_end <- realtime_end
  if (!is.null(limit)) q$limit <- limit
  if (!is.null(offset)) q$offset <- offset
  if (!is.null(sort_order)) q$sort_order <- sort_order
  if (!is.null(units)) q$units <- units
  if (!is.null(frequency)) q$frequency <- frequency
  if (!is.null(aggregation_method)) q$aggregation_method <- aggregation_method
  if (!is.null(output_type)) q$output_type <- output_type
  if (!is.null(vintage_dates)) q$vintage_dates <- vintage_dates

  res <- httr::GET(url, query = q, httr::accept("application/json"))
  if (httr::http_error(res)) stop(sprintf("FRED API request failed [%s]", httr::status_code(res)))
  txt <- httr::content(res, as = "text", encoding = "UTF-8")
  obj <- jsonlite::fromJSON(txt, flatten = TRUE)
  if (is.null(obj$observations)) stop("Estructura inesperada de la API de FRED; falta 'observations'.")
  return(obj$observations)
}


